src/charm.py

```diff
@@ -5,98 +5,161 @@
+from relations.opensearch import OpenSearchRelation
+from relations.postgresql import PostgresqlRelation
 from state import State
+from structured_config import CharmConfig
 
 logger = logging.getLogger(__name__)
 
 
-COMMANDS = {
-    "datahub-actions": "/bin/sh -c /start_datahub_actions.sh",
-    "datahub-frontend": "/bin/sh -c /start.sh",
-    "datahub-gms": "/bin/sh -c /datahub/datahub-gms/scripts/start.sh",
-    "datahub-mae-consumer": "/bin/sh -c /datahub/datahub-mae-consumer/scripts/start.sh",
-    "datahub-mce-consumer": "/bin/sh -c /datahub/datahub-mce-consumer/scripts/start.sh",
-}
+APPLICATIONS: List[applications.AbstractApplication] = [
+    applications.ActionsApplication,
+    applications.FrontendApplication,
+    applications.GMSApplication,
+    applications.MAEConsumerApplication,
+    applications.MCEConsumerApplication,
+]
 
 
-CONTAINERS = list(COMMANDS.keys())
-
-
-def get_pebble_layer(application_name, context):
+def get_pebble_layer(
+    application: applications.AbstractApplication, context: applications.ApplicationContext
+) -> Dict:
     """Create pebble layer based on application.
 
     Args:
-        application_name: Name of DataHub application.
-        context: Environment to include with the pebble plan.
+        application: Name of DataHub application.
+        context: Context from which to gather environment to include with the pebble plan.
 
     Returns:
         pebble plan dict.
     """
     return {
         "summary": "datahub layer",
         "services": {
-            application_name: {
-                "summary": application_name,
-                "command": COMMANDS[application_name],
-                "startup": "enabled",
+            application.name: {
+                "summary": application.name,
+                "command": application.command,
+                "startup": "enabled" if application.is_enabled(context) else "disabled",
                 "override": "replace",
                 # Included to make configuration changes replan the service.
-                "environment": context,
+                "environment": application.compile_environment(context),
             },
         },
     }
 
 
-class DatahubK8SOperatorCharm(ops.CharmBase):
+class DatahubK8SOperatorCharm(TypedCharmBase[CharmConfig]):
     """Charm the application."""
 
+    config_type = CharmConfig
+
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
         self._state = State(self.app, lambda: self.model.get_relation("peer"))
 
-        for container in CONTAINERS:
-            self.framework.observe(self.on[container].pebble_ready, self._on_pebble_ready)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.peer_relation_changed, self._on_peer_relation_changed)
+        # TODO (mertalpt): Decide whether to implement or remove this.
+        # self.framework.observe(self.on.update_status, self._on_update_status)
```

Comment:
This should be implemented in line with our other charms. In this hook, we validate that the container(s) has a valid pebble plan. If not, it re-creates it. We also perform health checks on the containers here before setting the charm status to active, if these exist for any of your application containers.